central/sensor/service/service_impl_test.go

```diff
@@ -55,3 +75,216 @@ func TestGetCertExpiryStatus(t *testing.T) {
 		})
 	}
 }
+
+// CRS Test Suite.
+
+type crsTestSuite struct {
+	suite.Suite
+	mockCtrl *gomock.Controller
+	context  context.Context
+}
+
+func (s *crsTestSuite) SetupTest() {
+	s.mockCtrl = gomock.NewController(s.T())
+	s.context = sac.WithAllAccess(context.Background())
+	utils.Should(testutils.LoadTestMTLSCerts(s.T()))
+}
+
+func TestCrs(t *testing.T) {
+	suite.Run(t, new(crsTestSuite))
+}
+
+func (s *crsTestSuite) TestCrsCentralReturnsAllServiceCertificates() {
+	sensorService, mockServer := newSensorService(s.context, s.mockCtrl)
+
+	// First-time CRS cluster registration.
+	mockServer.prepareNewHandshake(sensorHello)
+	err := sensorService.Communicate(mockServer)
+	s.NoError(err)
+	s.Len(mockServer.clustersRegistered, 1, "expected exactly one registered cluster")
+	centralHello := retrieveCentralHello(s, mockServer)
+	certBundle := centralHello.GetCertBundle()
+	s.Len(certBundle, 15, "expected 15 entries (1 CA cert, 7 service certs, 7 service keys) in bundle")
+}
+
+func (s *crsTestSuite) TestCrsFlowCanBeRepeated() {
+	sensorService, mockServer := newSensorService(s.context, s.mockCtrl)
+
+	// First-time CRS cluster registration.
+	mockServer.prepareNewHandshake(sensorHello)
+	err := sensorService.Communicate(mockServer)
+	s.NoError(err)
+	s.Len(mockServer.clustersRegistered, 1, "expected exactly one registered cluster")
+
+	// Initiating the CRS flow a second time should still work.
+	mockServer.prepareNewHandshake(sensorHello)
+	err = sensorService.Communicate(mockServer)
+	s.NoError(err)
+	s.Len(mockServer.clustersRegistered, 1, "expected exactly one registered cluster")
+	// Verify that we again got all certificates we need.
+	centralHello := retrieveCentralHello(s, mockServer)
+	certBundle := centralHello.GetCertBundle()
+	s.Len(certBundle, 15, "expected 15 entries (1 CA cert, 7 service certs, 7 service keys) in bundle")
+}
+
+func (s *crsTestSuite) TestCrsFlowFailsAfterLastContact() {
+	sensorService, mockServer := newSensorService(s.context, s.mockCtrl)
+
+	// CRS cluster registration.
+	mockServer.prepareNewHandshake(sensorHello)
+	err := sensorService.Communicate(mockServer)
+	s.NoError(err)
+
+	// Simulate a connection with service certificates has occurred by updating
+	// the LastContact field of a cluster.
+	mockServer.clustersRegistered[0].HealthStatus = &storage.ClusterHealthStatus{
+		LastContact: timestamppb.Now(),
+	}
+
+	// Initiating the CRS should fail now.
+	mockServer.prepareNewHandshake(sensorHello)
+	err = sensorService.Communicate(mockServer)
```

Comment:
Perhaps assert that the error message contains `forbidden to connect with a Cluster Registration Certificate as already-existing cluster` here?

Reply:
428625226c.